design-proposals/003-credentials.md

```diff
@@ -0,0 +1,226 @@
+
+Management of user and system credentials is also performed by credential provider, for instance initial setup of system credentials or refresh of user credentials.
+
+Credential provider consists of:
+ * **handler** - to implement OAuth flow (REST)
+ * **applicator** - to apply credentials to connection properties
+ * **deployer** - to augment k8s deployment descriptor
+ * **manager** - to manage credentials
+
+## Domain
+
+There are two types of domain objects, differentiated by security: public data for UI and sensitive secrets for the backend.
+
+Credentials need to be combined or linked together with other credentials, this has merits when considering that OAuth credentials are partialy owned by the user and partialy owned by the system. For example Syndesis owns non-public credentials (OAuth Client Secret, X.509 certificate and private key) and semi-public (OAuth Client ID), that need to be used during acquisition and in a running integration together with private credentials associated with the user (username, OAuth authorization token, OAuth refresh token). This makes for two credentials with two different security and lifecycle requirements.
+
+NOTE: Salesforce supports using x509 certificates for OAuth flow, and that should be preferred as the refreshing OAuth token can be performed within certificate's expiry i.e. not depending on the refresh token.
+
+### Credential view model
+
+Each credential should have a name so that the user can identify it, type: this can be username-password, x509 or OAuth tokens, internal identifier and connector label so it can be determined to what connection it can be applied to.
+
+The credential name can be auto generated (e.g. "Salesforce credentials for user@example.com").
+
+View model credential properties:
+
+ * **id**        : System assigned identifier for the credential, e.g. UUID
+ * **name**      : Name of the credential, for humans, e.g. "Salesforce credentials for user@example.com"
+ * **connector** : Connector type (label) that the credential applies to, e.g. "salesforce"
+ * **created**   : Date/time the credential was created
+ * **note**      : Optional note, could be generated during credential acquisition, e.g. "Expires 23.11.2018"
+ 
+This constitutes the view model, notice that the sensitive (secret) data is not exposed to the UI -- the UI should reference credentials by the identifier.
+
+The view model is generated by the backend from kubernetes secrets, mapping secret name to connection, attributes above as annotations of the secret.
+
+### Credential backend model
+
+As credential providers would share common implementation details, surely for application of credentials, i.e. creating the kubernetes secret from acquired data, this facilitates the need for a common credential backend model. This backend model can be viewed as data transfer object between credential provider, logic that handles the application of credentials and persistence in kubernetes secrets.
```

Comment:
The reload mechanism in [spring-cloud-kubernetes](https://github.com/fabric8io/spring-cloud-kubernetes/blob/master/spring-cloud-kubernetes-core/src/main/java/io/fabric8/spring/cloud/kubernetes/reload/ConfigReloadAutoConfiguration.java) supports reloading secrets also (it reloads the whole Spring ApplicationContext for matching change), and secrets are [already being used](https://github.com/syndesisio/syndesis-rest/blob/master/openshift/src/main/java/io/syndesis/openshift/OpenShiftServiceImpl.java#L208-L216); or am I missing something?
I think it makes sense to have separate models for connector properties that are stored in database and have lower security requirements vs credentials stored in k8s secrets having greater security requirements.